controllers/csi/daemonset.go

```diff
@@ -0,0 +1,356 @@
+					},
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{
+						driver,
+						registrar,
+						livenessprobe,
+					},
+					ServiceAccountName: DaemonSetName,
+					Volumes:            vol,
+				},
+			},
+		},
+	}
+
+	dsHash, err := generateDaemonSetHash(ds)
+	if err != nil {
+		return nil, err
+	}
+	ds.Annotations[statefulset.AnnotationTemplateHash] = dsHash
+
+	return ds, nil
+}
+
+func generateDaemonSetHash(ds *appsv1.DaemonSet) (string, error) {
+	data, err := json.Marshal(ds)
+	if err != nil {
+		return "", err
+	}
+
+	hasher := fnv.New32()
+	_, err = hasher.Write(data)
+	if err != nil {
+		return "", err
+	}
+
+	return strconv.FormatUint(uint64(hasher.Sum32()), 10), nil
+}
+
+func prepareMetadata() metav1.ObjectMeta {
+	return metav1.ObjectMeta{
+		Name:      DaemonSetName,
+		Namespace: "dynatrace",
+		Labels: map[string]string{
+			"dynatrace.com/operator": "dynatrace",
+		},
+		Annotations: map[string]string{},
+	}
+}
+
+func prepareDriverSpec(operatorImage string) v1.Container {
+	privileged := true
+	userID := int64(0)
+	bidirectional := v1.MountPropagationBidirectional
+
+	return v1.Container{
+		Name:    "driver",
+		Image:   operatorImage,
+		Command: []string{"csi-driver"},
+		Args: []string{
+			"--endpoint=unix://csi/csi.sock",
+			"--node-id=$(KUBE_NODE_NAME)",
+			"--health-probe-bind-address=:10080",
+		},
+		Env: []v1.EnvVar{
+			{
+				Name: "POD_NAMESPACE",
+				ValueFrom: &v1.EnvVarSource{
+					FieldRef: &v1.ObjectFieldSelector{
+						FieldPath: "metadata.namespace",
+					},
+				},
+			},
+			{
+				Name: "KUBE_NODE_NAME",
+				ValueFrom: &v1.EnvVarSource{
+					FieldRef: &v1.ObjectFieldSelector{
+						APIVersion: "v1",
+						FieldPath:  "spec.nodeName",
+					},
+				},
+			},
+		},
+		ImagePullPolicy: v1.PullAlways,
+		Ports: []v1.ContainerPort{
+			{
+				Name:          "healthz",
+				Protocol:      v1.ProtocolTCP,
+				ContainerPort: 10080,
+			},
+		},
+		LivenessProbe: &v1.Probe{
+			FailureThreshold:    3,
+			InitialDelaySeconds: 5,
+			PeriodSeconds:       5,
+			SuccessThreshold:    1,
+			TimeoutSeconds:      1,
+			Handler: v1.Handler{
+				HTTPGet: &v1.HTTPGetAction{
+					Path:   "/healthz",
+					Port:   intstr.FromString("healthz"),
+					Scheme: "HTTP",
+				},
+			},
+		},
+		SecurityContext: &v1.SecurityContext{
+			Privileged: &privileged,
+			RunAsUser:  &userID,
+			SELinuxOptions: &v1.SELinuxOptions{
+				Level: "s0",
+			},
+		},
+		VolumeMounts: []v1.VolumeMount{
+			{
+				Name:      "plugin-dir",
+				MountPath: "/csi",
+			},
+			{
+				Name:             "plugins-dir",
+				MountPath:        "/var/lib/kubelet/plugins",
+				MountPropagation: &bidirectional,
+			},
+			{
+				Name:             "mountpoint-dir",
+				MountPath:        "/var/lib/kubelet/pods",
+				MountPropagation: &bidirectional,
+			},
+			{
+				Name:             "dynatrace-oneagent-data-dir",
+				MountPath:        "/data",
+				MountPropagation: &bidirectional,
+			},
+		},
+	}
+}
```

Comment:
could some of this code also be moved to functions? e.g. a function to create the env vars, one for the volume mounts, etc. would make the function more readable and more concise